pr/upgrade and refactor

noahehall opened the pull request:
**What**: 
- typescript first
- react 18 & functional components
- bun.sh (tests, bundling, development)
- upgrade ~~waypoint~~ all dependencies

<!-- Why are these changes necessary? -->

**Why**:
- waypoint upgrade was required: check the growing list of stale issues (waypoint default export is now a named export)
- while im here might as well update them all

<!-- How were these changes implemented? -->

**How**: 
- no additional program changes were made, however the migration should inspire additional PRs to:
  - fully utilize react 18 features
  - drop remaining legacy react patterns and dependencies
- some tests were disabled  

<!-- Have you done all of these things?  -->

**Checklist**: 

<!-- add "N/A" to the end of each line that's irrelevant to your changes -->

<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

* [ ] Documentation
* [ ] Tests
* [ ] Ready to be merged <!-- In your opinion, is this ready to be merged as soon as it's reviewed? -->
* [ ] Added myself to contributors table <!-- this is optional, see the contributing guidelines for instructions -->
* [ ] find a bun compatible styleguidist alternative
* [ ] use framer to animate replacement of image

**feel free to add additional comments**:
- https://github.com/react-cosmos/rfcs/issues/27
- https://github.com/oven-sh/bun/issues/2877

ill try to finish this PR sometime next weekend, but verified it works
![image](https://github.com/oven-sh/bun/assets/10324554/6ed19fb4-f1c9-4483-9448-e034f316ae03)

noahehall committed:
Feat(core)!: refactor and upgrade

- remove non typescript-first dependencies
- upgrade all deps to latest versions
- refactor components to react.fc
- migrate to bun.sh

noahehall committed:
Feat(tests)!: migrate tests to bun:test

- skipped tests requiring dom: https://github.com/oven-sh/bun/issues/198

stereobooster commented on package.json:
```diff
@@ -77,5 +69,8 @@
-  "homepage": "https://github.com/stereobooster/react-ideal-image#readme"
+  "homepage": "https://github.com/stereobooster/react-ideal-image#readme",
+  "dependencies": {
+    "@ladle/react": "^2.12.3"
```
I think this meant to be in `devDependencies`

stereobooster commented on global.d.ts:
```diff
@@ -0,0 +1,153 @@
+
+declare global {
+  // @see https://stackoverflow.com/questions/38383676/error-ts2339-property-connection-does-not-exist-on-type-navigator
+  interface Navigator extends NavigatorNetworkInformation {}
```
I'm not sure about overriding built-in interfaces in npm packages. This is a normal approach in the "final" project, but in npm package this is questionable

stereobooster commented on src/components/IdealImage/index.tsx:
```diff
@@ -0,0 +1,352 @@
+    width,
+  }
+
+  console.info('\n\n rendering with props', useProps)
```
I guess this is for debugging

stereobooster commented:
I quickly got over it. I t would take some time to fully review

noahehall commented:
awesome; i don't think it will be ready to review for a week or two; but early recommendations would definitely be helpful

noahehall commented:
> I quickly got over it. I t would take some time to fully review

I would hold off on reviewing this until I remove the `Draft` label

noahehall committed:
Fix(MR): remove network typings and update prettier

noahehall committed:
Fix(MR): cleanup code smells

noahehall commented on src/components/IdealImage/index.tsx:
```diff
@@ -0,0 +1,282 @@
+    });
+    setImgState({
+      ...imgState,
+      state: loadState, // TODO(noah): rename to status
```
blah missed this one

noahehall commented on src/components/IdealImageWithDefaults/fixtures/freeimage3072×2304.jpg:
delete this file

noahehall commented on src/components/IdealImageWithDefaults/fixtures/index.stories.tsx:
```diff
@@ -0,0 +1,13 @@
+import IdealImage from "../../IdealImage";
+import img from "./freeimage3072×2304.jpg";
+
+export default <div>I am a fixture</div>;
```
try again with bun 0.6.3 - it's working for nirv

noahehall commented:
@stereobooster my main goals now are to:
- ensure i've checked all the checkboxes in the checklist
- ensure ive updated all the package json scripts ive removed
- react-cosmos

noahehall commented on src/components/IdealImageWithDefaults/fixtures/freeimage3072×2304.jpg:
delete this

noahehall commented on src/components/IdealImage/index.tsx:
```diff
@@ -0,0 +1,282 @@
+    if (
+      !imgState.url ||
+      ssr ||
+      [loadStates.loaded, loadStates.loaded].includes(imgState.state)
```
think this should be loading and loaded

noahehall commented on src/components/IdealImage/index.tsx:
```diff
@@ -0,0 +1,282 @@
+  });
+
+  const message = getMessage(icon, imgState);
+  const useProps = {
```
this pattern smells really bad

noahehall commented on src/components/Media/index.tsx:
```diff
@@ -0,0 +1,132 @@
+      <noscript>
+        <img
+          {...compose(useProps.theme.img, useProps.theme.noscript)}
+          src={useProps.src}
```
think these should use the nsSrc props from fallbackParams

noahehall commented on src/components/helpers.ts:
```diff
@@ -0,0 +1,145 @@
+  const first = notWebp[0];
+
+  // TODO(noah): required to pass test
+  if (!ssr || !first) return {};
```
check if this is still required

noahehall committed:
Chore(deps): install deps after upgrade

noahehall commented:
@stereobooster i've gotten the [develop branch](https://github.com/stereobooster/react-ideal-image/compare/master...noahehall:react-idealer-image:develop) to a stable state

However, i've opted to move in a  different direction than the existing code patterns, and additionally
- no clue when bun.sh will work with react.cosmos, and dont plan on dealing with storybook or the other non-bunnish frameworks
- all the tests are broken (i might have time to fix in month or so)
- dropped support for legacy browsers
- replaced waypoint with framer-motion

If, however, you've taken a look at my develop branch and would still consider merging into your repo, i would totally support that (vs continuing on my fork)

let me know! 
cheers 

- nh
